apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -538,28 +539,35 @@ void init(List<LifecycleListener> lifecycleListeners) {
         }
     }
 
-    public synchronized void start() {
-        if (getConfig(CoreConfiguration.class).getDelayInitMs() > 0) {
-            startWithDelay();
+    public synchronized void start(boolean premain) {
+        long delayInitMs = getConfig(CoreConfiguration.class).getDelayInitMs();
+        if (premain && shouldDelay()) {
+            delayInitMs = Math.max(delayInitMs, 5000L);
+        }
+        if (delayInitMs > 0) {
+            startWithDelay(delayInitMs);
         } else {
             startSync();
         }
     }
 
-    private synchronized void startWithDelay() {
+    private boolean shouldDelay() {
+        return ClassLoader.getSystemClassLoader().getResource("org/apache/catalina/startup/Bootstrap.class") != null;
+    }
+
+    private synchronized void startWithDelay(final long delayInitMs) {
         ThreadPoolExecutor pool = ExecutorUtils.createSingleThreadDeamonPool("tracer-initializer", 1);
         pool.submit(new Runnable() {
             @Override
             public void run() {
                 try {
-                    long delayInitMs = coreConfiguration.getDelayInitMs();
                     logger.info("Delaying initialization of tracer for " + delayInitMs + "ms");
```

Comment:
Can we end the delay once the platform MBean server is initialized (`!MBeanServerFactory.findMBeanServer(null).isEmpty()`)? In that case, it should be guaranteed that JUL is initialized.

Reply:
You mean poll every 500ms or so? It still doesn't guarantee anything.
I think we can live with the 5 seconds delay in Tomcat on 1.8 installed via premain

Reply:
You'd only wait up to 500ms more than necessary vs waiting up to 5s more than necessary and possibly not waiting long enough. We could use the same mechanism for `delay_init` and the MBean server polling. Both conditions need to be fulfilled to end the delay.

> I think we can live with the 5 seconds delay in Tomcat on 1.8 installed via premain

I think so too, just mentioning for completeness